source/jsonParsing.ml

```diff
@@ -7,60 +7,280 @@
+    exception
+      TypeError of {
+        message: string;
+        json: t;
+      }
+
+    exception
+      ParseException of {
+        message: string;
+        location: Location.t;
+      }
+
+    let null_node = { Node.location = Location.null_location; value = `Null }
+
+    let json_of_string input =
+      let decode decoder =
+        match Jsonm.decode decoder with
+        | `Lexeme lexeme -> lexeme
+        | `Error error ->
+            raise
+              (ParseException
+                 {
+                   message = Format.asprintf "%a" Jsonm.pp_error error;
+                   location = Location.from_decoded_range (Jsonm.decoded_range decoder);
+                 })
+        | `End
+        | `Await ->
+            raise
+              (ParseException
+                 {
+                   message = "Encountered end of file or input";
+                   location = Location.from_decoded_range (Jsonm.decoded_range decoder);
+                 })
+      in
+      let rec parse_value value post_parse_function decoder =
+        let location = Location.from_decoded_range (Jsonm.decoded_range decoder) in
+        match value with
+        | `Os -> parse_object [] post_parse_function decoder
+        | `As -> parse_array [] post_parse_function decoder
+        | (`Null | `Bool _ | `String _ | `Float _) as value ->
+            post_parse_function { Node.location; value } decoder
+        | _ ->
+            raise (ParseException { message = "Encountered unexpected token or element"; location })
+      and parse_array current_values post_parse_function decoder =
+        let location = Location.from_decoded_range (Jsonm.decoded_range decoder) in
+        match decode decoder with
+        | `Ae ->
+            post_parse_function { Node.value = `List (List.rev current_values); location } decoder
+        | element ->
+            parse_value
+              element
+              (fun value -> parse_array (value :: current_values) post_parse_function)
+              decoder
+      and parse_object current_children post_parse_function decoder =
+        let location = Location.from_decoded_range (Jsonm.decoded_range decoder) in
+        match decode decoder with
+        | `Oe ->
+            post_parse_function
+              { Node.value = `Assoc (List.rev current_children); location }
+              decoder
+        | `Name name ->
+            parse_value
+              (decode decoder)
+              (fun value -> parse_object ((name, value) :: current_children) post_parse_function)
+              decoder
+        | _ ->
+            raise (ParseException { message = "Encountered unexpected token or element"; location })
+      in
+      let decoder = Jsonm.decoder (`String input) in
+      try
+        let json = parse_value (decode decoder) (fun v _ -> v) decoder in
+        Result.Ok json
+      with
+      | ParseException { message; location } ->
+          Result.Error (Error.create ~kind:ParseError ~location message)
```

Comment:
Since we just convert the exception into an error, what about exposing two variants of this:

```
val from_string_exn : string -> t (* throws ParseException *)

val from_string : string -> (t, ParseError.t) Result.t
```

And the second one can be implemented with the first one, using your code:
```suggestion
    let from_string string =
      try
        Result.Ok (from_string_exn string)
      with
      | ParseException { message; location } ->
          Result.Error (Error.create ~kind:ParseError ~location message)
```

Reply:
Much much better! Thanks!